File: py/fitter.py
from iminuit import Minuit

from driver import LL

import numpy as np

class LLFit(object):
    ll = LL(1., 0., 0.6, -0.6, 0.)
    
    def __init__(self):
        pass
        
    @staticmethod
    def fcn(alpha, dphi, alph1, alph2, xi):
        LLFit.ll.set(alpha, dphi, alph1, alph2, xi)
        loglh = LLFit.loglh()
        print('loglh: {}, a {}, phi {}, a1 {}, a2 {}, xi {}'.format(
            loglh, alpha, dphi, alph1, alph2, xi))
        return loglh
    
    @staticmethod
    def loglh():
        return -np.sum(np.log(LLFit.ll(LLFit.data))) +\
            np.log(np.sum(LLFit.ll(LLFit.normdata))) * len(LLFit.data)

    @staticmethod
    def fitTo(data, normdata, init=None):
        LLFit.data = data
        LLFit.normdata = normdata
        
        minimizer = Minuit(
            LLFit.fcn,
            alpha= 0.6,       error_alpha=0.1, limit_alpha=(-1., 1.),       fix_alpha=False,
             dphi= 0.5*np.pi, error_dphi =1.0, limit_dphi =(-np.pi, np.pi), fix_dphi =False,
            alph1= 0.6,       error_alph1=0.1, limit_alph1=(0., 1.),        fix_alph1=False,
            alph2=-0.6,       error_alph2=0.1, limit_alph2=(-1., 0.),       fix_alph2=False,
               xi= 0.0,       error_xi   =0.5, limit_xi   =(-1.01, 1.01),   fix_xi   =False,
            errordef=0.5)
        
        fmin, param = minimizer.migrad()
        minimizer.print_param()
        return (fmin, param)

File: src/fitter.cpp
#include "fitter.h"

#include "Minuit2/MnMigrad.h"

using namespace ROOT::Minuit2;
using namespace llfit;

// 0.6, 0.5*3.1415, 0.6, -0.6
Fitter::Fitter(TheFCN& fcn) : m_fcn(fcn) {
    m_upar.Add("alpha", 0.6, 0.1, -1., 1.);
    m_upar.Add("dPhi", 1.55, 0.1, -3.1415, 3.1415);
    m_upar.Add("alpha1", 0.6, 0.1, -1., 1.);
    m_upar.Add("alpha2", -0.6, 0.1, -1., 1.);
    m_upar.Add("xi", 0.8, 0.1, -1.05, 1.05);
}

FunctionMinimum Fitter::fit() {
    MnMigrad migrad(m_fcn, m_upar);
    return migrad();
}
